File: 01000/900/1955_수식_표현/BJ_1955_수식_표현.py
# https://www.acmicpc.net/problem/1955

import math
import sys


readline = lambda: sys.stdin.readline().rstrip()

n = int(readline())
factorials = [math.factorial(i) for i in range(8)]

dp = [float('inf')] * (n + 1)
dp[1] = 1

for k in range(2, n + 1):
    for i in range(1, k):
        dp[k] = min(dp[k], dp[i] + dp[k - i])
    for i in range(1, math.floor(math.sqrt(k)) + 1):
        dp[k] = min(dp[k], dp[i] + dp[k // i]) if k % i == 0 else dp[k]
    for i in range(2, min(k, len(factorials))):
        dp[k] = min(dp[k], dp[i]) if k == factorials[i] else dp[k]

print(dp[n])

File: 01000/900/1955_수식_표현/BJ_1955_수식_표현.go
// https://www.acmicpc.net/problem/1955

package main

import (
	"bufio"
	"fmt"
	"os"
	"runtime"
)

type Fin struct {
	*bufio.Reader
}

func (in Fin) ReadInt() (n int) {
	const GOOS_IS_WINDOWS = (runtime.GOOS == "windows")

	for {
		b, err := in.ReadByte()
		if b < '0' || '9' < b || err != nil {
			if GOOS_IS_WINDOWS && b == '\r' {
				in.ReadByte()
			}
			return n
		}

		n = n*10 + int(b-'0')
	}
}

func Solve(goal int) {
	visited := make([]bool, 10_001)
	visited[1] = true

	var dp []map[int]any
	dp = append(dp, map[int]any{1: nil})

	visitFactorial := func(k int, callee any) {
		if k < 8 {
			n := 1
			for 0 < k {
				n *= k
				k--
			}

			if !visited[n] {
				visited[n] = true
				dp[len(dp)-1][n] = nil
				callee.(func(int, any))(n, callee)
			}
		}
	}

	for !visited[goal] {
		dp = append(dp, map[int]any{})

		for i := 0; i < len(dp)-1; i++ {
			for a := range dp[i] {
				for b := range dp[(len(dp)-1)-(i+1)] {
					if a+b < len(visited) && !visited[a+b] {
						visited[a+b] = true
						dp[len(dp)-1][a+b] = nil
						visitFactorial(a+b, visitFactorial)
					}
					if a*b < len(visited) && !visited[a*b] {
						visited[a*b] = true
						dp[len(dp)-1][a*b] = nil
						visitFactorial(a*b, visitFactorial)
					}
				}
			}
		}
	}

	fmt.Printf("%d\n", len(dp))
}

func main() {
	Solve(Fin{bufio.NewReader(os.Stdin)}.ReadInt())
}
